pkg/util/pki/parse.go

```diff
@@ -140,3 +140,161 @@ func DecodeX509CertificateRequestBytes(csrBytes []byte) (*x509.CertificateReques
 
 	return csr, nil
 }
+
+type PEMBundle struct {
+	CAPEM    []byte
+	ChainPEM []byte
+}
+
+type chainNode struct {
+	cert *x509.Certificate
+
+	issuer *chainNode
+}
+
+// ParseCertificateChainPEM will decode the bundle before passing to
+// ParseCertificateChain
+func ParseCertificateChainPEM(pembundle []byte) (*PEMBundle, error) {
+	certs, err := DecodeX509CertificateChainBytes(pembundle)
+	if err != nil {
+		return nil, err
+	}
+	return ParseCertificateChain(certs)
+}
+
+// ParseCertificateChain will return an x509 PEM encoded certificate chain and
+// CA. The CA may not be a true root, but the highest intermediate certificate.
+// Removes duplicate entries. The returned CA may be empty if a single
+// certificate was passed. Chain contains leaf certificate first. Removes
+// comments and white space. An error is returned if the passed bundle is not a
+// valid flat tree chain, the bundle is malformed, or the chain is broken.
+func ParseCertificateChain(certs []*x509.Certificate) (*PEMBundle, error) {
+	// De-duplicate certificates
+	for i := 0; i < len(certs)-1; i++ {
+		for j := 1; j < len(certs); j++ {
+			if i == j {
+				continue
+			}
+			if certs[i].Equal(certs[j]) {
+				certs = append(certs[:i], certs[i+1:]...)
+			}
+		}
+	}
+
+	var chains []*chainNode
+	// Build the nodes to the linked list
+	for i := range certs {
+		chains = append(chains, &chainNode{cert: certs[i]})
+	}
```

Comment:
Just to add onto this also. By methodically building the linked list in this way we also prove 2 things:
- The chain itself is valid, in that each certificate passes `CheckSignatureFrom` for some other certificate in the chain
- The chain is complete in that there is no breaks in the chain

Reply:
Thank you so much for the comment you added, it helps a lot!!

> By methodically building the linked list in this way we also prove 3 things

Totally agree, keeping those 3 invariants while iterating is a sure way of making sure the invariant still holds on the resulting `chain[0]`.

Reply:
> It sounds like this becomes a breaking change then for our issuers - previously, they were able to provide a 'broken' chain, and it's up to clients to either handle that (or not) accordingly.

This function is only used for the internal cert-manager issuers. External issuers are still free to populate `status.CA` and `status.Certificate` however they like. 

> I think we should be very careful with a change that adds this sort of enforcement. Do we need this in order to fix the root CA handling issue? I envisioned that we'd just add a simple check that looks for a self signed root, and adds it to ca.crt if it is found.

It was my understanding that the original issue was to solve the problem where the `status.CA` may include an intermediate, rather than a true root. When discussed, I understood we wanted to cater to both use cases. This can only be done AFAIK by constructing the full chain. Can a broken chain be present for the ACME/CA/Vault/Venafi Issuers?

Reply:
> It sounds like this becomes a breaking change then for our issuers - previously, they were able to provide a 'broken' chain, and it's up to clients to either handle that (or not) accordingly.

This function is only used for the internal cert-manager issuers. External issuers are still free to populate `status.CA` and `status.Certificate` however they like. 

> I think we should be very careful with a change that adds this sort of enforcement. Do we need this in order to fix the root CA handling issue? I envisioned that we'd just add a simple check that looks for a self signed root, and adds it to ca.crt if it is found.

It was my understanding that the original issue was to solve the problem where the `status.CA` may include an intermediate, rather than a true root. When discussed, I understood we wanted to cater to both use cases. This can only be done AFAIK by constructing the full chain. Can a broken chain be present for the ACME/CA/Vault/Venafi Issuers?